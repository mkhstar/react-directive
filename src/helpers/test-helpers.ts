import { ReactElement } from "react";
// eslint-disable-next-line import/no-extraneous-dependencies
import renderer from "react-test-renderer";

export const snapshotTest = (
  render: ReactElement<any, string>,
  title?: string
) => {
  const tree = renderer.create(render).toJSON();
  expect(tree).toMatchSnapshot(title || "");
};

export const truthyValues = ["value", true, 1, Infinity, -1, -Infinity, {}, []];
export const falsyValues = ["", false, 0, null];
export const truthyFalsyMap = truthyValues
  .map((value) =>
    falsyValues.map((falsyValue) => ({ truthy: value, falsy: falsyValue }))
  )
  .flat();
