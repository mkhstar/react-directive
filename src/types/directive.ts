import { makeDirective } from "../directives/makeDirective";
import { Classes, ClassesDependencies } from "./classes";

export type DirectiveChildren<T> =
  | ((currentItem: T, index: number) => React.ReactNode)
  | React.ReactNode;

export interface DirectiveProps<T = undefined, E = undefined> {
  dirIf?: any;
  dirShow?: any;
  className?: Classes;
  classNameDeps?: ClassesDependencies;
  dirFor?: Array<T>;
  dirRef?: React.Ref<
    E extends keyof HTMLElementTagNameMap ? HTMLElementTagNameMap[E] : any
  >;
  children?: T extends undefined ? React.ReactNode : DirectiveChildren<T>;
  keyExtractor?: T extends undefined
    ? never
    : (currentItem: T) => string | number | null;
}

export type DirectiveMap = {
  [Key in keyof JSX.IntrinsicElements]: ReturnType<typeof makeDirective<Key>>;
};
