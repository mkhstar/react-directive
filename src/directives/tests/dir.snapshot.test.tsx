import React from "react";
import directive from "..";
import { snapshotTest } from "../../helpers/test-helpers";

describe("anchor element", () => {
  describe("renderes anchor elements correctly", () => {
    it("Anchor with href and title", () => {
      snapshotTest(
        <directive.a href="https://github.com/mkhstar" title="anchor" />
      );
    });
    it("Anchor with dirShow set to false", () => {
      snapshotTest(
        <directive.a dirShow={false}>Some text content</directive.a>
      );
    });
    it("Anchor with dirIf set to false", () => {
      snapshotTest(<directive.a dirIf={false}>Some text content</directive.a>);
    });

    it("Anchor with classNames", () => {
      snapshotTest(
        <directive.a
          className={{ "should-show": true, "should-not-show": false }}
        >
          Some text content
        </directive.a>
      );
    });
  });
});

describe("div element", () => {
  it("renders correctly for div with no properties", () => {
    snapshotTest(<directive.div />);
  });

  it("renders correctly for div with dirIf set to false", () => {
    snapshotTest(<directive.div dirIf={false} />);
  });

  it("renders correctly for div with dirShow set to false", () => {
    snapshotTest(<directive.div dirShow={false} />);
  });

  it("renders correctly for div with className", () => {
    snapshotTest(<directive.div className="test-class" />);
  });

  it("renders correctly for div with class generated by useClassName hook", () => {
    snapshotTest(<directive.div className={{ test: true }} />);
  });

  it("renders correctly for div with dirFor", () => {
    snapshotTest(
      <directive.div dirFor={[1, 2, 3]}>{(item) => item}</directive.div>
    );
  });
});
