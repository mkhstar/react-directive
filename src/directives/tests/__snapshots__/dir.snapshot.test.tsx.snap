// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`anchor element renderes anchor elements correctly Anchor with classNames 1`] = `
<a
  className="should-show"
>
  Some text content
</a>
`;

exports[`anchor element renderes anchor elements correctly Anchor with dirIf set to false 1`] = `null`;

exports[`anchor element renderes anchor elements correctly Anchor with dirShow set to false 1`] = `
<a
  style={
    {
      "display": "none",
    }
  }
>
  Some text content
</a>
`;

exports[`anchor element renderes anchor elements correctly Anchor with href and title 1`] = `
<a
  href="https://github.com/mkhstar"
  title="anchor"
/>
`;

exports[`div element renders correctly for div with class generated by useClassName hook 1`] = `
<div
  className="test"
/>
`;

exports[`div element renders correctly for div with className 1`] = `
<div
  className="test-class"
/>
`;

exports[`div element renders correctly for div with dirFor 1`] = `
[
  <div>
    1
  </div>,
  <div>
    2
  </div>,
  <div>
    3
  </div>,
]
`;

exports[`div element renders correctly for div with dirIf set to false 1`] = `null`;

exports[`div element renders correctly for div with dirShow set to false 1`] = `
<div
  style={
    {
      "display": "none",
    }
  }
/>
`;

exports[`div element renders correctly for div with no properties 1`] = `<div />`;
